pkg/controller/gkenetworkparamset/gkenetworkparamset_controller.go

```diff
@@ -174,27 +209,93 @@ func (c *Controller) syncGKENetworkParamSet(ctx context.Context, key string) err
 	}
 
 	params := obj.(*networkv1.GKENetworkParamSet)
+	originalParams := params.DeepCopy()
 
-	// TODO: Enable finalizer addition when finalizer deletion is added.
-	// addFinalizerToGKENetworkParamSet(params)
-	// update will be done once in deferred function call
-	// if err := c.updateGKENetworkParamSet(ctx, params); err != nil {
-	// 	return err
-	// }
+	defer func() {
+		if !reflect.DeepEqual(originalParams, params) {
+			if updateErr := c.updateGKENetworkParamSet(ctx, params); updateErr != nil {
+				err = multierror.Append(updateErr, err)
+			}
+		}
+
+		if !reflect.DeepEqual(originalParams.Status, params.Status) {
+			if updateErr := c.updateGKENetworkParamSetStatus(ctx, params); updateErr != nil {
+				err = multierror.Append(updateErr, err)
+			}
+		}
+	}()
+
+	if params.DeletionTimestamp != nil {
+		// GKENetworkParamSet is being deleted, handle the delete event
+		return c.handleGKENetworkParamSetDelete(ctx, params)
+	}
 
-	subnet, err := c.gceCloud.GetSubnetwork(c.gceCloud.Region(), params.Spec.VPCSubnet)
+	addFinalizerInPlace(params)
+
+	subnet, subnetValidation := c.getAndValidateSubnet(ctx, params)
+	params.Status.Conditions = updateConditions(params.Status.Conditions, subnetValidation.toCondition())
+	if !subnetValidation.IsValid {
+		return
+	}
+
+	paramsValidation, err := c.validateGKENetworkParamSet(ctx, params, subnet)
 	if err != nil {
-		fetchSubnetErrs.Inc()
-		return err
+		return
+	}
+	params.Status.Conditions = updateConditions(params.Status.Conditions, paramsValidation.toCondition())
+	if !paramsValidation.IsValid {
+		return
```

Comment:
ditto

Reply:
see above comment, I dont think we need to log validation results since theyre going to be placed in the object's conditions